File: post_processing/pm25/grimm_aggregate.R
library('parallel')

CDepict = function(c){
  v <- c('P','C0:','C0;','c0:','c0;','C1:','C1;','c1:','c1;','C2:','C2;','c2:','c2;','C3:','C3;','c3:','c3;','C4:','C4;','c4:','c4;','C5:','C5;','c5:','c5;','C6:','C6;','c6:','c6;','C7:','C7;','c7:','c7;','C8:','C8;','c8:','c8;','C9:','C9;','c9:','c9;')
  id <- which(v == c)
  return(id)
}

Clean = function(tbl){
  vec <- c(NA,'P','C0:','C0;','c0:','c0;','C1:','C1;','c1:','c1;','C2:','C2;','c2:','c2;','C3:','C3;','c3:','c3;','C4:','C4;','c4:','c4;','C5:','C5;','c5:','c5;','C6:','C6;','c6:','c6;','C7:','C7;','c7:','c7;','C8:','C8;','c8:','c8;','C9:','C9;','c9:','c9;')
  t2 <- tbl[,1]
  index <- c()
  for (i in 1:(length(t2)-1)){
    if (!(CDepict(t2[i])-CDepict(t2[i+1])) %in% c(-1,40)){
      index <- c(index,i)
      i_1 <- i
      i_2 <- i
      while (t2[i_1] != 'c9;'){
        index <- c(index,i_1)
        i_1 <- i_1-1
      }
      while (t2[i_2] != 'P'){
        index <- c(index,i_2)
        i_2 <- i_2+1
      }
    }
  }
  if (t2[length(t2)] != "c9;"){
    a <- length(t2) 
    while (t2[a] != "c9;"){
      index <- c(index,a)
      a <- a-1
    }
  }
  index <- unique(index)
  nfmins <- length(index)
  if (!identical(index, c())){
    rmv <- tbl[-c(index),]
  } else{
    rmv <- tbl
  }
  rows <- nrow(rmv)
  rownames(rmv) <- c(1:rows)
  
  mtemp <- data.frame(matrix(NA, nrow=rows*5/41, ncol=17))
  
  mtemp$X1 <- rep(unlist(c('P','C_:','C_;','c_:','c_;')))
  for (i in 1:(rows/41)){
    for (j in 1:17){
      mtemp[((i-1)*5+1),j] <- rmv[(1+(i-1)*41),j]
    }
    for (j in 2:10){
      for (k in 2:5){
        mtemp[((i-1)*5+k),j]<- 10*mean(as.numeric(c(rmv[((i-1)*41+k),j],rmv[((i-1)*41+k+4),j],rmv[(((i-1)*41+k+2*4)),j],rmv[(((i-1)*41+k+3*4)),j],rmv[(((i-1)*41+k+4*4)),j],rmv[(((i-1)*41+k+5*4)),j],rmv[(((i-1)*41+k+6*4)),j],rmv[(((i-1)*41+k+7*4)),j],rmv[(((i-1)*41+k+8*4)),j],rmv[(((i-1)*41+k+9*4)),j])),na.rm = T)
      }
    }
  }
  for (i in 1:(nrow(mtemp)/5)){
    mtemp$X10[5*i]<-0
  }
  mtemp$X10[mtemp$X10 == 1600]<-160 
  mtemp$X10[is.nan(mtemp$X10)]<-NA 
  return(list(mtemp,nfmins))
  
  
}

SixtoMin_Parallel = function(path){
  cores <- detectCores(logical = 12) # EDW ALLAZEIS TON ARITHMO TWN PYRHNWN POY THELEIS NA XRHSIMOPOIHSEIS #
  table1 <- read.table(as.character(path),fill=T)
  vec <- c(NA,'P','C0:','C0;','c0:','c0;','C1:','C1;','c1:','c1;','C2:','C2;','c2:','c2;','C3:','C3;','c3:','c3;','C4:','C4;','c4:','c4;','C5:','C5;','c5:','c5;','C6:','C6;','c6:','c6;','C7:','C7;','c7:','c7;','C8:','C8;','c8:','c8;','C9:','C9;','c9:','c9;')
  if (!identical(which(!(table1[,1] %in% vec)),integer(0))){
    table <- table1[-c(which(!(table1[,1] %in% vec))),]
    nslines <- length(c(which(!(table1[,1] %in% vec))))
  } else {
    table <- table1
    nslines <- 0
  }
  rows1 <- nrow(table)
  rownames(table) <- c(1:rows1)
  t1 <- table[,1]
  
  cuts <- c()
  sep_points <- c()
  parts <- list()
  for (i in 1:(cores-1)){
    cuts[i] <- ceiling(rows1*(i/cores))
    sep_points[i] <- which(t1[cuts[i]:rows1] == "P")[1]
    sep_points[i] <- cuts[i] + sep_points[i]-1
  }
  sep_points <- c(1,sep_points,rows1)
  for (i in sep_points[1:(length(sep_points)-1)]){
    parts[[which(sep_points == i)]] <- table[sep_points[which(sep_points == i)]:(sep_points[which(sep_points == i)+1]-1),]
  }
  
  clust <- makeCluster(cores) 
  clusterExport(clust, 'CDepict')
  parts1 <- parLapply(clust ,parts[1:(length(sep_points)-1)] ,Clean)
  final <- data.frame()
  nfmins1 <- 0
  for (i in 1:length(parts1)){
    final <- rbind(final,parts1[[i]][[1]])
    nfmins1 <- nfmins1+parts1[[i]][[2]]
  } 
  
  print("Number of non-standard lines found and removed: ") 
  print(nslines)
  print("Number of lines of uncompleted minutes found and removed:") 
  print(nfmins1)
  return(final)
}

Final <- SixtoMin_Parallel("input.grimm")

write.table(Final, "output.csv", sep=",", quote = FALSE, row.names = FALSE, col.names = FALSE, na="" )

File: post_processing/pm25/grimm_aggregate.py
import pandas as pd
import numpy as np
from multiprocessing import Pool, cpu_count

def CDepict(c):
    v = ['P', 'C0:', 'C0;', 'c0:', 'c0;', 'C1:', 'C1;', 'c1:', 'c1;', 'C2:', 'C2;', 'c2:', 'c2;', 
         'C3:', 'C3;', 'c3:', 'c3;', 'C4:', 'C4;', 'c4:', 'c4;', 'C5:', 'C5;', 'c5:', 'c5;', 
         'C6:', 'C6;', 'c6:', 'c6;', 'C7:', 'C7;', 'c7:', 'c7;', 'C8:', 'C8;', 'c8:', 'c8;', 
         'C9:', 'C9;', 'c9:', 'c9;']
    return v.index(c)


table1 = pd.read_csv( "temp-input.grimm",
                      header=None, na_filter=False, names=range(17), low_memory=False )
vec = [ np.nan, 'P',
        'C0:','C0;','c0:','c0;', 'C1:','C1;','c1:','c1;',
        'C2:','C2;','c2:','c2;', 'C3:','C3;','c3:','c3;',
        'C4:','C4;','c4:','c4;', 'C5:','C5;','c5:','c5;',
        'C6:','C6;','c6:','c6;', 'C7:','C7;','c7:','c7;',
        'C8:','C8;','c8:','c8;','C9:','C9;','c9:','c9;' ]

if not table1.iloc[:, 0].isin(vec).all():
    tbl = table1[table1.iloc[:, 0].isin(vec)]
    nslines = len(table1) - len(tbl)
else:
    tbl = table1
    nslines = 0

bad_idx = []
prev_idx = None
for i in range(len(tbl.index)-1):
    idx0 = tbl.index[i]
    idx1 = tbl.index[i+1]
    dd = CDepict(tbl.loc[idx0,0]) - CDepict(tbl.loc[idx1,0])
    if (dd != -1) and (dd != 40):
        print( f"Bad {idx0} ; {tbl.loc[idx0,0]} {CDepict(tbl.loc[idx0,0])} ; {tbl.loc[idx1,0]} {CDepict(tbl.loc[idx1,0])}" )
        bad_idx.append(idx0)
        j = i
        while (j>=0) and (tbl.loc[tbl.index[j],0] != 'c9;'):
            bad_idx.append(tbl.index[j])
            j -= 1
        j = i
        while (j < len(tbl)) and (tbl.loc[tbl.index[j],0] != 'P'):
            bad_idx.append(tbl.index[j])
            j += 1

#if t2[-1] != "c9;":
#    a = len(t2) - 1
#    while t2[a] != "c9;":
#        bad_idx.append(a)
#        a -= 1

bad_idx = list(set(bad_idx))

nfmins = len(bad_idx)
print( f"Number of lines of uncompleted minutes removed: {nfmins}" )

rmv = tbl.drop(bad_idx) if bad_idx else tbl
rows = len(rmv)
rmv.index = range(1, rows + 1)

# Initialize dataframe with -1.
# Negative values never appear, and we must remove
# unwritten the cells with post-editing of the CSV
mtemp = pd.DataFrame(-1, index=range(int(rows * 5 / 41)), columns=range(17) )

# The first column is the times/channels characters
mtemp[0] = ['P', 'C_:', 'C_;', 'c_:', 'c_;'] * (rows // 41)

for i in range(rows // 41):
    mtemp.iloc[i * 5, 0] = str( rmv.iloc[i * 41, 0] )
    # Copy the times lines (P lines)
    for j in range(1,17):
        mtemp.iloc[i * 5, j] = int( rmv.iloc[i * 41, j] )
    # Aggregate the channels (C lines)
    for j in range(1, 10):
        for k in range(1, 5):
            ll = [rmv.iloc[i * 41 + k, j], rmv.iloc[i * 41 + k + 4, j],
                  rmv.iloc[i * 41 + k + 2 * 4, j], rmv.iloc[i * 41 + k + 3 * 4, j],
                  rmv.iloc[i * 41 + k + 4 * 4, j], rmv.iloc[i * 41 + k + 5 * 4, j],
                  rmv.iloc[i * 41 + k + 6 * 4, j], rmv.iloc[i * 41 + k + 7 * 4, j],
                  rmv.iloc[i * 41 + k + 8 * 4, j], rmv.iloc[i * 41 + k + 9 * 4, j]]
            vv = 0
            for l in ll:
                try:
                    vv += int(l)
                    mtemp.iloc[i * 5 + k, j] = vv
                except:
                    if l != "": print( f"Funny value: {l}" )
        # end for k in range(1, 5)
    # end for j in range(1, 10)


for i in range(len(mtemp) // 5):
    mtemp.iloc[5 * i - 1, 9] = 0

mtemp.iloc[:, 9] = mtemp.iloc[:, 9].replace(1600, 160)
#mtemp.iloc[:, 9] = mtemp.iloc[:, 9].replace(np.nan, np.nan)

mtemp.to_csv( "temp-aggregated.csv", sep=",", header=False, index=False, na_rep="" )
